File: solutions/2026-01-29/cpp/2026-01-29-minimum-cost-to-convert-string-i.cpp
```cpp
// Problem: Minimum Cost to Convert String I
// Link: https://leetcode.com/problems/minimum-cost-to-convert-string-i/
// Approach:
// This problem can be modeled as a shortest path problem on a graph.
// Each lowercase English letter ('a' through 'z') can be considered a node in the graph.
// The given `original`, `changed`, and `cost` arrays define directed edges between these character nodes.
// An edge exists from `original[i]` to `changed[i]` with a weight of `cost[i]`.
// Since we want the minimum cost to change one character to another, and there might be multiple paths (e.g., 'a' -> 'c' -> 'b' might be cheaper than 'a' -> 'b' directly),
// we need to find the shortest path between each pair of characters that need to be converted.
// The Floyd-Warshall algorithm is suitable here because the number of nodes (characters) is small (26), and it computes all-pairs shortest paths.
// We initialize a 26x26 distance matrix `dist` where `dist[i][j]` stores the minimum cost to convert the i-th character to the j-th character.
// Initially, `dist[i][i]` is 0 for all i, and `dist[i][j]` is infinity for i != j.
// Then, we populate the `dist` matrix with direct conversion costs from the input arrays. If multiple direct conversions exist for the same pair, we take the minimum cost.
// After populating with direct costs, we run the Floyd-Warshall algorithm:
// For k from 0 to 25:
//   For i from 0 to 25:
//     For j from 0 to 25:
//       dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
// This ensures that `dist[i][j]` holds the minimum cost to convert the i-th character to the j-th character, considering all intermediate conversions.
// Finally, we iterate through the `source` and `target` strings. If `source[i]` is different from `target[i]`, we add the `dist[source[i] - 'a'][target[i] - 'a']` to the total cost.
// If at any point `dist[source[i] - 'a'][target[i] - 'a']` is still infinity, it means the conversion is impossible, and we return -1.
//
// Time Complexity:
// 1. Initializing the distance matrix: O(26*26) = O(1) since the alphabet size is constant.
// 2. Populating direct costs: O(N) where N is the length of `original`, `changed`, and `cost`.
// 3. Floyd-Warshall algorithm: O(26^3) = O(1) since the alphabet size is constant.
// 4. Iterating through source and target strings: O(L) where L is the length of `source` and `target`.
// Therefore, the dominant factor is O(L).
//
// Space Complexity:
// O(26*26) = O(1) for the distance matrix, as the alphabet size is constant.
//
class Solution {
public:
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        // Initialize a 26x26 matrix to store minimum costs between any two characters.
        // Each cell dist[i][j] will store the minimum cost to convert the i-th character to the j-th character.
        // We use long long to avoid potential overflow with large costs.
        // Initialize with a large value representing infinity.
        const long long INF = 1e18; // A sufficiently large number for infinity
        vector<vector<long long>> dist(26, vector<long long>(26, INF));

        // The cost to convert a character to itself is 0.
        for (int i = 0; i < 26; ++i) {
            dist[i][i] = 0;
        }

        // Populate the distance matrix with the direct conversion costs given in the input.
        // If there are multiple direct paths between two characters, we take the minimum cost.
        for (int i = 0; i < original.size(); ++i) {
            int u = original[i] - 'a';
            int v = changed[i] - 'a';
            dist[u][v] = min(dist[u][v], (long long)cost[i]);
        }

        // Apply the Floyd-Warshall algorithm to find the shortest path between all pairs of characters.
        // This accounts for indirect conversions (e.g., 'a' -> 'c' -> 'b' might be cheaper than 'a' -> 'b').
        for (int k = 0; k < 26; ++k) { // Intermediate node
            for (int i = 0; i < 26; ++i) { // Starting node
                for (int j = 0; j < 26; ++j) { // Ending node
                    // If a path from i to k and a path from k to j exist,
                    // check if going through k provides a shorter path from i to j.
                    if (dist[i][k] != INF && dist[k][j] != INF) {
                        dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
                    }
                }
            }
        }

        // Calculate the total minimum cost to convert the source string to the target string.
        long long totalCost = 0;
        for (int i = 0; i < source.length(); ++i) {
            int u = source[i] - 'a';
            int v = target[i] - 'a';

            // If the characters at the current position are different, we need to convert.
            if (u != v) {
                // Get the minimum cost to convert source[i] to target[i].
                long long conversionCost = dist[u][v];

                // If conversionCost is still INF, it means it's impossible to convert
                // the character at this position, so return -1.
                if (conversionCost == INF) {
                    return -1;
                }
                // Add the conversion cost to the total cost.
                totalCost += conversionCost;
            }
        }

        // Return the total minimum cost.
        return totalCost;
    }
};
```

File: solutions/2026-01-29/python/2026-01-29-minimum-cost-to-convert-string-i.py
```python
import heapq

# Problem: Minimum Cost to Convert String I
# Link: https://leetcode.com/problems/minimum-cost-to-convert-string-i/
#
# Approach:
# This problem can be modeled as finding the shortest path in a graph.
# Each lowercase English letter ('a' through 'z') can be considered a node in the graph.
# The given `original`, `changed`, and `cost` arrays define the directed edges and their weights.
# An entry `original[i] = u`, `changed[i] = v`, `cost[i] = w` means there's a directed edge
# from node `u` to node `v` with weight `w`.
# Since we can perform operations multiple times, we might need to find indirect paths (e.g.,
# changing 'a' to 'c' then 'c' to 'b' might be cheaper than a direct 'a' to 'b' conversion).
# Therefore, we need to find the minimum cost to convert any character `u` to any other
# character `v`. This is a classic All-Pairs Shortest Path problem.
# We can use Dijkstra's algorithm starting from each character to find the minimum cost
# to reach all other characters. Since there are only 26 possible characters, we can run
# Dijkstra 26 times, which is efficient enough.
#
# The steps are:
# 1. Build an adjacency list representation of the graph where `graph[u]` stores pairs `(cost, v)`
#    representing an edge from `u` to `v` with `cost`.
# 2. For each character `i` from 'a' to 'z':
#    a. Run Dijkstra's algorithm starting from `i` to find the minimum cost to reach every other
#       character. Store these minimum costs in a 2D array `min_costs[start_char_code][end_char_code]`.
#       Initialize `min_costs` with infinity.
# 3. Iterate through the `source` and `target` strings. For each index `k`:
#    a. If `source[k]` is already equal to `target[k]`, the cost is 0.
#    b. Otherwise, find the minimum cost to convert `source[k]` to `target[k]` using the
#       precomputed `min_costs` array. Add this cost to the total cost.
#    c. If `min_costs[source[k]][target[k]]` is still infinity, it means the conversion is
#       impossible, return -1.
# 4. Return the total accumulated cost.
#
# Time Complexity Analysis:
# - Building the graph: O(E), where E is the number of conversion rules (cost.length).
# - Running Dijkstra's for each character:
#   - There are 26 possible source characters.
#   - For each source character, Dijkstra's algorithm on a graph with V nodes (26 letters)
#     and at most E edges (or E directed edges from each node if we consider all possibilities,
#     but practically it's bounded by the input rules) using a min-heap takes O(E_prime * log V),
#     where E_prime is the number of edges explored from a source.
#     In our case, V = 26. The number of edges from any node is at most 26 (to other letters).
#     However, the input `cost` array can have up to 2000 entries, so `E_prime` can be up to 2000.
#     So, for one source character, it's roughly O(2000 * log 26).
#   - Total for all 26 source characters: 26 * O(E * log V) = 26 * O(2000 * log 26).
#   - A tighter bound considering the number of edges is O(V * (E + V log V)) if E is dense,
#     but here V is constant (26). So it's O(26 * (E + 26 log 26)). Given E <= 2000, this is efficient.
#     A more precise analysis for Dijkstra with E edges on V vertices is O(E log V).
#     Since V=26 is constant, it's O(E). Running this 26 times is O(26 * E).
#     However, if we are careful about graph construction, the total edges from all nodes could be up to 2000.
#     The number of unique edges is at most 26*26.
#     The total number of edges in the graph is `cost.length`.
#     Dijkstra for each node: 26 * (E_total_edges + V log V) -> 26 * (2000 + 26 log 26). This is effectively O(26 * E).
# - Calculating total cost: O(N), where N is the length of source/target strings.
#
# Overall Time Complexity: O(26 * E + N) which is dominated by O(E) for precomputation if E > N, or O(N) if N > E.
# Given E <= 2000 and N <= 10^5, it's O(N + E).
#
# Space Complexity Analysis:
# - Adjacency list for the graph: O(26 * 26) or O(E) depending on implementation. Max edges is 2000.
#   Since we store edges for each character, it's O(E) where E is `cost.length`.
# - `min_costs` matrix: O(26 * 26) = O(1) because the alphabet size is constant.
# - Dijkstra's priority queue and visited set: O(V) = O(26) = O(1).
#
# Overall Space Complexity: O(E) due to the adjacency list storing conversion rules.

def minimumCost(source: str, target: str, original: list[str], changed: list[str], cost: list[int]) -> int:
    # Initialize a 2D array to store the minimum cost to convert character i to character j.
    # The size is 26x26, representing 'a' through 'z'.
    # Initialize all costs to infinity, except for self-conversion which is 0.
    # Using a large number to represent infinity.
    INF = float('inf')
    min_costs = [[INF] * 26 for _ in range(26)]

    # Set the cost of converting a character to itself to 0.
    for i in range(26):
        min_costs[i][i] = 0

    # Populate the min_costs matrix with direct conversion costs.
    # If there are multiple ways to convert char_o to char_c, take the minimum cost.
    for i in range(len(original)):
        u = ord(original[i]) - ord('a')
        v = ord(changed[i]) - ord('a')
        c = cost[i]
        min_costs[u][v] = min(min_costs[u][v], c)

    # Floyd-Warshall algorithm to find all-pairs shortest paths.
    # This is an alternative to running Dijkstra 26 times and might be simpler to implement here
    # given the small number of vertices (26).
    # For each intermediate character k, check if going from i to k then k to j is cheaper
    # than the current direct path from i to j.
    for k in range(26):  # Intermediate character
        for i in range(26):  # Start character
            for j in range(26):  # End character
                if min_costs[i][k] != INF and min_costs[k][j] != INF:
                    min_costs[i][j] = min(min_costs[i][j], min_costs[i][k] + min_costs[k][j])

    total_cost = 0
    n = len(source)

    # Iterate through the source and target strings to calculate the total conversion cost.
    for i in range(n):
        src_char_code = ord(source[i]) - ord('a')
        tgt_char_code = ord(target[i]) - ord('a')

        # If the characters are already the same, the cost is 0 for this position.
        if src_char_code == tgt_char_code:
            continue

        # Get the minimum cost to convert source[i] to target[i] from the precomputed matrix.
        conversion_cost = min_costs[src_char_code][tgt_char_code]

        # If the conversion cost is still infinity, it means it's impossible to convert.
        if conversion_cost == INF:
            return -1

        # Add the conversion cost for this character to the total cost.
        total_cost += conversion_cost

    # Return the minimum total cost to convert the entire source string to the target string.
    return total_cost

```